## Hacker News Project

The purpose is to analyze the data set and answer the following two questions: 
* Find out if Ask HN or Show HN posts recive more comments on average
* Do posts created at a certain time receive more comments on average?

In [7]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn = hn[1:]
print(headers)
print(hn[:5])


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [8]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1].lower()
    if title.startswith('ask hn'):
        ask_posts.append(title)
    elif title.startswith('show hn'):
            show_posts.append(title)
    else: 
        other_posts.append(title)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
print(ask_posts[:2])

1744
1162
17194
['ask hn: how to improve my personal website?', 'ask hn: am i the only one outraged by twitter shutting down share counts?']


In [9]:
print(ask_posts[:5])
print('\n')
print(show_posts[:5])

['ask hn: how to improve my personal website?', 'ask hn: am i the only one outraged by twitter shutting down share counts?', 'ask hn: aby recent changes to css that broke mobile?', 'ask hn: looking for employee #3 how do i do it?', 'ask hn: someone offered to buy my browser extension from me. what now?']


['show hn: wio link  esp8266 based web of things hardware development platform', 'show hn: something pointless i made', 'show hn: shanhu.io, a programming playground powered by e8vm', 'show hn: webscope  easy way for web developers to communicate with clients', 'show hn: geoscreenshot  easily test geo-ip based web pages']


In [10]:
print(headers)

total_ask_comments = 0
for comments in ask_posts:
    num_comments = row[4]
    num_comments = int(num_comments)
    total_ask_comments += num_comments
    avg_ask_comments = total_ask_comments / len(ask_posts)
    
print(total_ask_comments)
print(avg_ask_comments)

total_show_comments = 0
for comments in show_posts:
    num_comments = row[4]
    num_comments = int(num_comments)
    total_show_comments += num_comments
    avg_show_comments = total_show_comments / len(ask_posts)
    
print(total_show_comments)
print(avg_show_comments)



['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
101152
58.0
67396
38.64449541284404


Ask Posts receive more comments than Show Posts.  
* Average Ask Post Comments = 58
* Average Show Post Comments = 38.6

In [11]:
print(ask_posts[:1])
print('\n')
print(hn[:1])

['ask hn: how to improve my personal website?']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']]


In [12]:
import datetime as dt

result_list = []
for row in ask_posts:
    created_at = row[6]
    n_comments = row[4]
    result_list.append([created_at, n_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    time = row[0]
    date_time = dt.datetime.strptime(time, "%m/%d/%Y %H:%M")
    hour = date_time.strftime("%H")
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = int(row[1])
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += int(row[1])

ValueError: time data ':' does not match format '%m/%d/%Y %H:%M'

Number of comments is column 4
Date is column 6
Date Example: '1/26/2016 19:30'
dt.datetime.strptime(date_1_str, "%m %d, %Y")